### DISCUSSION TITLE
Avalonia.FuncUIを使うべきかAvaloniaだけで使うべきか

### DISCUSSION THREAD

**Discussion by murata2makoto** [fffsharp/Forum#8] [雑談]
第五回Fun Fan F#でAvalonia.FuncUIを知りました（有難うございます）。使おうと思って調べているのですが、Avaloniaを素で使うほうが良いのではと思いつつあります。

まず、私はF#に魂を売ってますが、XMLにはそれ以上です（元W3C XML WGメンバ）。XAMLを見るにはなんの抵抗もありません。

Visual StudioのAvalonia拡張をインストールするとF#用のAvalonia利用プロジェクトが作れますし、XAMLの画面レイアウトを確認しながら編集出来ます。わざわざF#で書く理由はXAMLを見たくない以外にはあるのでしょうか？

**Comment by fffsharp** [comment]
3 秒で思いついたのは、UI も型レベルで扱えるとかですかね？+ MVU の仕掛けを食い込ませるのに都合が良いとかが、パッと思いつきました。

**Reply by murata2makoto** [reply]
確かにXAMLで書いただけじゃ型にはなりませんね。https://github.com/fsprojects/FsXaml なんていうのもありますが、Avalonia用ではなく本家MS用です。

**Reply by fffsharp** [reply]
FsXaml はチラ見した感じでは、WPF 用っぽいので、Avalonia では使えなさそうな感じがしますね。
// ただ、試してはいないので、真偽不明ですが…

**Reply by murata2makoto** [reply]
名前空間がMSのものになっていてAvalonia用じゃないですからAvaloniaには使えません。

**Comment by SilkyFowl** [comment]
関数合成や部分適用+ジェネリクスを使えるのが大きいですね。
たとえばこういう関数を定義して……

```fs
let validatedTextBox (domain: Domain<string, _, _>) value addAttrs dispatchSetValueMsg =
    let invalidTextAttrs text errors =
        [ TextBox.text text
          List.map box errors |> TextBox.errors ]

    TextBox.create [
        match value with
        | Valid v -> domain.ToDto v |> TextBox.text
        | Invalid (CreateFailed (dto, errors)) -> yield! invalidTextAttrs dto errors
        | Invalid (UpdateFailed (ValueNone, dto, errors)) -> yield! invalidTextAttrs dto errors
        | Invalid (UpdateFailed (ValueSome before, dto, errors)) ->
            yield! invalidTextAttrs dto errors
            TextBox.onLostFocus (fun _ -> (domain.ToDto >> dispatchSetValueMsg) before)
        | Invalid (MargedError marged) -> List.map box marged |> TextBox.errors

        yield! addAttrs

        TextBox.onTextChanged dispatchSetValueMsg
    ]

let pathSelectorView domain value (buttonText: string) buttonCallback dispatchSetValueMsg addAttrs =

    Grid.create [
        Grid.rowDefinitions "Auto,28"
        Grid.columnDefinitions "Auto,*"
        Grid.margin (0, 0, 4, 0)
        yield! addAttrs
        Grid.children [
            Button.create [
                Button.column 0
                Button.row 0
                Button.content buttonText
                Button.onClick buttonCallback
            ]
            validatedTextBox
                domain
                value
                [ TextBox.column 1
                  TextBox.row 0
                  TextBox.rowSpan 2
                  TextBox.verticalAlignment VerticalAlignment.Top ]
                dispatchSetValueMsg
        ]
    ]
```

こう使います。

```fs
let playListFilePathView model dispatch =
    Component.create (
        "playListFilePath-View",
        fun ctx ->

            let value =
                model
                |> ctx.useReadMapCurrent (fun m -> m.Settings.Settings.PlayListFilePath)

            let buttonCallback _ =
                (PickPlayList >> SettingsMsg) Started |> dispatch

            let dispatchSetValueMsg s =
                (SetPlayListFilePath >> SettingsMsg) s |> dispatch

            ctx.attrs [ Grid.column 0 ]
            pathSelectorView playListFilePath value "PlayList" buttonCallback dispatchSetValueMsg []
    )

let snapShotFolderPathView model dispatch =
    Component.create (
        "snapShotFolderPath-View",
        fun ctx ->
            let value =
                model
                |> ctx.useReadMapCurrent (fun m -> m.Settings.Settings.SnapShotFolderPath)

            let buttonCallback _ =
                (PickSnapshotFolder >> SettingsMsg) Started
                |> dispatch

            let dispatchSetValueMsg s =
                (SetSnapShotFolderPath >> SettingsMsg) s
                |> dispatch

            ctx.attrs [ Grid.column 1 ]

            pathSelectorView snapShotFolderPath value "SnapShotFolder" buttonCallback dispatchSetValueMsg []
    )
```

**Reply by murata2makoto** [reply]
これはAvaonia.FuncUIを使ったときだけ可能で、AvaloniaをF#から使ったときは不可能なんでしょうか？

**Reply by SilkyFowl** [reply]
はい、`Control.create [attrs]`というUI記述関数などはAvaonia.FuncUI固有です。

**Reply by SilkyFowl** [reply]
ただ、FuncUIはXAMLのようにリアルタイムで画面を確認する機能は未だ存在しません。
私はXAMLの支援が貧弱なvscodeを使っているのでデメリットにはなりませんでしたが、Visual Studioを使うならAvaloniaだけ(≒XAMLを使う)ほうがいい気がします。

現状をどうにかしようとLivePreviewerを開発中ですが、まだ最低限の動作が出来上がったところです。
配布可能なレベルまで完成したら......

https://user-images.githubusercontent.com/16532218/169697054-bda20cb1-852a-4aa0-bc1b-79e6adeffe61.mp4

**Reply by murata2makoto** [reply]
そう、XAMLでプレビュー出来るのは大きなメリットです。一方、F#のメリットを生かすのはやはりFuncUIのようですね。情報ありがとうございます。LivePreviewer期待してます。

**Comment by murata2makoto** [comment]
素でAvaloniaをF#から使うという方針でAvaloniaを調べています。

プログラムを実行することなく、XAMLから得られた画面レイアウトをプレビューできるというのは、やはり巨大なメリットです。ここが何とかならない限りFuncUIは使い物にならないでしょうね。そして、component方式は、その可能性を捨てているように思えます。

いっぽう、F#からAvaloniaを使うための説明資料の少なさに四苦八苦しています。そして、F#のオブジェクト指向機能は使わないことにしてきたのでプロパティの作り方も知りませんでした。

それでも、何とかViewModelを作ることができました。やっていることは、Smalltalk-80のpluggable viewとあんまり変わらないような気がしますし、bindingなどは制約型プログラミングの一部を入れているような感じ。今後は、ずっとAvaloniaでGUIを書くことになりそうです。（以前、C#で書いたら死にそうになった。）

**Reply by SilkyFowl** [reply]
ご苦労さまです……

私の経験上、DeepLを頼りにAvalonia関連のリポジトリのIssueやDiscussion、Stackoverflowを巡回するのが良いかなと思います……

ViewModelでしたらF#フレンドリーなEpoxyが第一選択肢に上がると思います
今もメンテされている上に、国産なので日本語のDocがバッチリ完備されています
https://github.com/kekyo/Epoxy/blob/main/README.ja.md

作者であるkekyo氏の知見も大変参考になるのでおすすめです
https://www.kekyo.net/

ところで、FuncUIのLivePreviewのv0.0.1-alphaをそろそろリリースできるかもしれません
https://github.com/SilkyFowl/Avalonia.FuncUI.LiveView/pull/1

**Reply by murata2makoto** [reply]
ありがとうございます。

意外と困るのが、Visual Studioの思わぬ動作です。code-behindにバグがあるとXAMLのプレビューが動かなくなるのも分からなくて悩みました。アイコンファイルに日本語名を使っているとなぜか動作しません。

**Comment by murata2makoto** [comment]
最近、またAvaloniaでGUIを作っています。

XAML+コードビハインドで書いて400行を超えたので、コードビハインドはやめようと思います。type X with ....でファイルを分けてあがいてみましたが、分けたファイル間で自由に相互参照できません。

MVVMで数日やってみましたが、これはC#で使うもんだと痛感しました。F#だとコードが冗長になって耐えられません。F#でもコードを簡単にする方法としてReactiveUI.Fodyを使うという手がありますが、コンパイルがとても困難なので諦めました。人に渡すかもしれないプログラムでFodyを使うのは絶対やめておくほうが良いでしょうね。

結局、FuncUIかElmishにしようと方針を変えました。XAML+コードビハインドで書いたものを、ChatGPTでMVUに書き直してもらいました。しかし、コンパイルできなかったり、画面が真っ白になったりして、思うようにいかなくてあがいています。

ChatGPTやCoPilotは使いまくりです。もちろん助かりますが、嘘も結構つきますね。

**Comment by murata2makoto** [comment]
結局、Elmishにしました。ずいぶん苦労しました。

ボタンとメッセージの対応、メッセージの連鎖については、CoPilotでグラフをdot記法で書いてSVGを生成して文書化しています。

**Reply by murata2makoto** [reply]
ずっと作っていたものを金曜にデモし、YouTube動画で公開しています。
[読みに困難のある子どもたちに公平なテストをーーアクセシブルCBTとPNPの可能性](https://www.youtube.com/live/OQ7DijF5cVI?si=UwglaeY7EOmyDuf-&t=2981)

MVUで書き、Avalonia.Controls.NativeWebViewを使ってEdgeでのプレビューを組み込んでます。これには[Avalonia Accelerate](https://avaloniaui.net/accelerate)にライセンス料金を払う必要がありました。日本円で1万5000円ぐらいです。

Avalonia.Controls.NativeWebViewを使わない方法もありますが、その場合はChromiumを組み込んだ巨大アプリになります。これは嫌がられたので諦めました。

**Reply by murata2makoto** [reply]
F# + Avalonia + MVU + OOXML + QTI3というプログラムになっています。以前、ここで発表したアウトラインアルゴリズムの改良版も使っています。オープンソースにすることが出来たら、またここでお知らせします。
